pkg/controller/mongodb/mongodb_controller.go

```diff
@@ -265,24 +265,46 @@ func (r ReplicaSetReconciler) ensureAutomationConfig(mdb mdbv1.MongoDB) error {
 	if err != nil {
 		return err
 	}
+
 	if err := r.client.CreateOrUpdate(&cm); err != nil {
 		return err
 	}
 	return nil
 }
 
-func buildAutomationConfig(mdb mdbv1.MongoDB, mdbVersionConfig automationconfig.MongoDbVersionConfig) automationconfig.AutomationConfig {
+func buildAutomationConfig(mdb mdbv1.MongoDB, mdbVersionConfig automationconfig.MongoDbVersionConfig, currentAc automationconfig.AutomationConfig) (automationconfig.AutomationConfig, error) {
 	domain := getDomain(mdb.ServiceName(), mdb.Namespace, "")
-	return automationconfig.NewBuilder().
+	newAc := automationconfig.NewBuilder().
 		SetTopology(automationconfig.ReplicaSetTopology).
 		SetName(mdb.Name).
 		SetDomain(domain).
 		SetMembers(mdb.Spec.Members).
 		SetMongoDBVersion(mdb.Spec.Version).
-		SetAutomationConfigVersion(1). // TODO: Correctly set the version
+		SetAutomationConfigVersion(currentAc.Version).
 		SetFCV(mdb.GetFCV()).
 		AddVersion(mdbVersionConfig).
 		Build()
+
+	// Here we compare the bytes of the two automationconfigs,
+	// we can't use reflect.DeepEqual() as it treats nil entries as different from empty ones,
+	// and in the AutomationConfig Struct we use omitempty to set empty field to nil
+	// The agent requires the nil value we provide, otherwise the agent attempts to configure authentication.
+
+	newAcBytes, err := json.Marshal(newAc)
+	if err != nil {
+		return automationconfig.AutomationConfig{}, err
+	}
+
+	currentAcBytes, err := json.Marshal(currentAc)
+	if err != nil {
+		return automationconfig.AutomationConfig{}, err
+	}
+
+	if bytes.Compare(newAcBytes, currentAcBytes) != 0 {
+		newAc.Version += 1
```

Comment:
another approach would be the builder to take in the previous version and upon calling build decides if it bumps the number: thus the AA comparisons will happen inside the AA builder code which is more relevant

Reply:
Would it make more sense to get rid of the function `setAutomationConfigVersion()` and replace it with a function that passes the previous version of the AC? Since the version should only be bumped in case of changes, this would prevent from setting the versions manually in unwanted ways. It would be set automatically to the one of the previous AC (0 if none is passed) and increased if something changed

Reply:
yes that what i think

Reply:
and it's the build method that would call getCurrentAutomationConfig